contracts/cw-proposal-multiple/src/proposal.rs

```diff
@@ -619,16 +675,174 @@ mod tests {
             votes.clone(),
             Uint128::new(13),
             true,
+            false,
         );
 
         // Should pass if majority voted
         assert!(prop.is_passed(&env.block).unwrap());
         assert!(!prop.is_rejected(&env.block).unwrap());
 
-        let prop = create_proposal(&env.block, voting_strategy, votes, Uint128::new(14), true);
+        let prop = create_proposal(
+            &env.block,
+            voting_strategy,
+            votes,
+            Uint128::new(14),
+            true,
+            false,
+        );
 
         // Shouldn't pass if only half voted
         assert!(!prop.is_passed(&env.block).unwrap());
         assert!(prop.is_rejected(&env.block).unwrap());
     }
+
+    #[test]
+    fn test_majority_revote_pass() {
+        // Revoting being allowed means that proposals may not be
+        // passed or rejected before they expire.
+        let env = mock_env();
+        let voting_strategy = VotingStrategy::SingleChoice {
+            quorum: voting::threshold::PercentageThreshold::Majority {},
+        };
+        let votes = MultipleChoiceVotes {
+            vote_weights: vec![Uint128::new(6), Uint128::new(0), Uint128::new(0)],
+        };
+
+        let prop = create_proposal(
+            &env.block,
+            voting_strategy.clone(),
+            votes.clone(),
+            Uint128::new(10),
+            false,
+            true,
+        );
+        // Quorum reached, but proposal is still active => no pass
+        assert!(!prop.is_passed(&env.block).unwrap());
+
+        let prop = create_proposal(
+            &env.block,
+            voting_strategy,
+            votes,
+            Uint128::new(10),
+            true,
+            true,
+        );
+        // Quorum reached & proposal has expired => pass
+        assert!(prop.is_passed(&env.block).unwrap());
+    }
+
+    #[test]
+    fn test_majority_revote_rejection() {
+        // Revoting being allowed means that proposals may not be
+        // passed or rejected before they expire.
+        let env = mock_env();
+        let voting_strategy = VotingStrategy::SingleChoice {
+            quorum: voting::threshold::PercentageThreshold::Majority {},
+        };
+        let votes = MultipleChoiceVotes {
+            vote_weights: vec![Uint128::new(5), Uint128::new(0), Uint128::new(0)],
+        };
+
+        let prop = create_proposal(
+            &env.block,
+            voting_strategy.clone(),
+            votes.clone(),
+            Uint128::new(10),
+            false,
+            true,
+        );
+        // No quorum reached, but proposal is still active => no rejection
+        assert!(!prop.is_rejected(&env.block).unwrap());
```

Comment:
Feel as if we may be able to make this test a little more targeted. How about constructing a situation where there is a tie and the total voting power has voted (five votes for first option and five for the second). Then we can verify that this case doesn't get hit:

https://github.com/DA0-DA0/dao-contracts/blob/cef33570e41c52b04acd692cdd527c836185fd13/contracts/cw-proposal-multiple/src/proposal.rs#L140-L144

Reply:
Sounds good, I made it revolve around a tie case with some checks now